src/CaptureFile/CaptureFile.cpp

```diff
@@ -431,6 +447,141 @@ ErrorMessageOr<void> CaptureFileImpl::WriteSectionList(
   return outcome::success();
 }
 
+ErrorMessageOr<uint64_t> CaptureFileImpl::CalculateContentEnd() const {
+  // If no section list exists, the end of the file is used.
+  if (header_.section_list_offset == 0) {
+    return GetEndOfFileOffset(fd_);
+  }
+
+  // If a user data section exists, then it has to be the last section.
+  if (const std::optional<uint64_t> user_data_section_index =
+          FindSectionByType(kSectionTypeUserData);
+      user_data_section_index.has_value()) {
+    if (user_data_section_index.value() != (section_list_.size() - 1)) {
+      return ErrorMessage{
+          "Unable to calculate where the content of the capture file ends: The user data section "
+          "is not the last section."};
+    }
+    return section_list_.back().offset + section_list_.back().size;
+  }
+
+  // Otherwise the section list is the last thing in the capture file.
+
+  if (IsThereSectionWithOffsetAfterSectionList()) {
+    return ErrorMessage{
+        "Unable to calculate where the content of the capture file ends: The file contains a non "
+        "user data section after the section list."};
+  }
+
+  return header_.section_list_offset + CalculateSectionListSizeInFile(section_list_.size());
+}
+
+ErrorMessageOr<void> CaptureFileImpl::VerifyCaptureFileValid() const {
+  if (FindAllSectionsByType(kSectionTypeUserData).size() > 1) {
+    return ErrorMessage{
+        "Capture file is invalid, because it contains more than 1 user data section."};
+  }
+
+  if (const std::optional<uint64_t> user_data_section_index =
+          FindSectionByType(kSectionTypeUserData);
+      user_data_section_index.has_value()) {
+    if (user_data_section_index.value() != section_list_.size() - 1) {
+      return ErrorMessage{
+          "Capture file is invalid, because the user data section is not the last section."};
+    }
+  }
```

Comment:
Nit: How about having a separated function for `VerifyUserDataSectionIsLastSection` since we use it multiple places

Reply:
Good idea! I solved this by introducing `ErrorMessageOr<bool> ContainsValidUserDataSection() const;`